### PR TITLE
fix(build): use cldr.version for cldr

### LINKED ISSUE #316: github action should use the cldr.version for checkout
```
        <cldr.version>0.0.0-SNAPSHOT-273d736f11</cldr.version>
```

- we should use for example `273d736f11` instead of checking out CLDR latest. Otherwise we could get a CLDR that is a mismatch with the code used.

### PR DESCRIPTION
**Pull request by srl295**
- github action should use the cldr.version for checkout
- also bump the version of some actions

Fixes #316

### CODE REVIEW

File: .github/workflows/build-jsp.yml [hunk 1]
```diff
@@ -11,12 +15,16 @@ jobs:
   build:
     runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@v2
+    - uses: actions/checkout@v3
+    - name: Get the CLDR_REF from pom.xml
+      id: cldr_ref
+      run: pip install yq && echo "CLDR_REF="$(xq -r '.project.properties["cldr.version"]' < pom.xml | cut -d- -f3) >> $GITHUB_OUTPUT && cat ${GITHUB_OUTPUT}
```
Comment (echeran) [inline .github/workflows/build-jsp.yml:21]:
```suggestion
      run: echo "CLDR_REF="$(mvn help:evaluate -Dexpression=cldr.version -q -DforceStdout | cut -d- -f3) >> $GITHUB_OUTPUT && cat ${GITHUB_OUTPUT}
```
I think you can replace the usage of `yq` with the Maven command that gives you the same.  Do you need the `$GITHUB_OUTPUT`, or can you remove that on the assumption that whatever is streamed to standard output will be captured just the same?

Comment (srl295) [reply .github/workflows/build-jsp.yml:21]: GITHUB_OUTPUT is for copying the value to another step in the process, so yes needed.

There's a maven command? I thought there might be but couldn't find it.

Comment (srl295) [reply .github/workflows/build-jsp.yml]: https://docs.github.com/en/actions/learn-github-actions/contexts#example-usage-of-the-steps-context

Comment (srl295) [reply .github/workflows/build-jsp.yml]: ok now i see it. VS Code showed me your comment but not the suggestion.

**Comment by srl295** [comment]
yes i will propagate this to other workflows once its…working

File: .github/workflows/push-jsp-on-tag.yml [hunk 2]
```diff
@@ -7,17 +7,22 @@ jobs:
   build-and-push-to-gcr:
     runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@v2
+    - uses: actions/checkout@v3
     - name: Set up JDK 11
       uses: actions/setup-java@v1
       with:
         java-version: 11
-    - uses: actions/checkout@v2
```
Comment (srl295) [inline .github/workflows/push-jsp-on-tag.yml:15]: yes, it was checking out the repo twice!

Comment (srl295) [reply .github/workflows/push-jsp-on-tag.yml:15]: probably a merge conflict

**Comment by srl295** [comment]
so a major issue here is that github checkout, following security best practices, no longer handles the short hash to CLDR -  https://github.com/actions/checkout/issues/265

so the short hash such as 273d736f11 in 0.0.0-SNAPSHOT-273d736f11 is a challenge to checkout.

I'm going back to an older version of the action, but it will be a stopgap.

**Comment by srl295** [comment]
well this turned out to be kind of a mess… see if the caching helps at all.

**Comment by srl295** [comment]
@echeran  remainng build errors seem unrelated - ideas?

**Comment by echeran** [comment]
> @echeran remainng build errors seem unrelated - ideas?

I think the second `mvn` command for the "Generate Confusables" tool execution should have `...-am -pl ...`, but I didn't add the `-am` part there.  Can you try adding that in to the corresponding lines in the failing workflows?

That seems like the only thing that is failing, so hopefully that is the fix needed to get this PR ready to merge.

**Comment by markusicu** [comment]
@srl295 please try @echeran 's suggestions. It would be great to get this PR done and in.

**Comment by srl295** [comment]
> > @echeran remainng build errors seem unrelated - ideas?
> 
> I think the second `mvn` command for the "Generate Confusables" tool execution should have `...-am -pl ...`, but I didn't add the `-am` part there. Can you try adding that in to the corresponding lines in the failing workflows?
> 
> That seems like the only thing that is failing, so hopefully that is the fix needed to get this PR ready to merge.

That was in the out-of-source, but the in-source was failing tests also.

File: .github/workflows/build-jsp.yml [hunk 3]
```diff
@@ -11,15 +15,31 @@ jobs:
...
+        key: cldr-${{ steps.cldr_ref.outputs.CLDR_REF }}
+        restore-keys: |
+          cldr
     - name: Check out CLDR
-      uses: actions/checkout@v2
+      uses: actions/checkout@v3
       with:
         repository: unicode-org/cldr
         path: cldr
+        ref: main
```
[the remark this answers is not in the document]
Comment (srl295) [reply .github/workflows/build-jsp.yml:36]: doesn't work because it's not a full hash.

**Comment by srl295** [comment]
OK, spot the problem with in-source build:

```shell
$ mkdir cldr
$ ln -s cldr ..
$ cd ../cldr
```

… fixed

**Comment by srl295** [comment]
@echeran @markusicu  in-source and python are passing again. out of source is getting much further. PTAL

Also rebased.

**Review by markusicu — APPROVED** [review]
looks... eminently plausible. thank you very much for getting this to work!

File: .flake8 [hunk 4]
```diff
@@ -5,6 +5,9 @@ statistics = True
...
 max-complexity = 10
-ignore = E302, # expected 2 blank lines, found 1
-         E701, # multiple statements on one line
-         W504, # line break after binary operator
+ignore = E302,
+       	 # expected 2 blank lines, found 1
+         E701,
+       	 # multiple statements on one line
+         W504,
+       	 # line break after binary operator
```
[the remark this answers is not in the document]
Comment (josh-hadley) [reply .flake8:13]: This has been biting tons of Python devs who use flake8 for linting. Inline  comments have always been "wrong", they're just enforcing it now  :-)

[the remark this answers is not in the document]
Comment (srl295) [reply .flake8:13]: a better error message would have been nice…

And yes, it's open source and i haven't opened a PR yet.

**Pull request merged**